src/backend/tests/kernelprog/vecadd/vecadd_macros.py

```diff
@@ -0,0 +1,123 @@
+# from numba_drvapi import (
+#     API_PROTOTYPES,
+#     cu_device
+#     cu_context
+#     cu_module
+#     cu_jit_option
+#     cu_jit_input_type
+#     cu_function
+#     cu_device_ptr
+#     cu_stream
+# )
+import numba_drvapi
+from ctypes import (
+    c_byte,
+    c_char_p,
+    # c_float,
+    c_int,
+    # c_size_t,
+    c_uint,
+    # c_uint8,
+    c_void_p,
+    # py_object,
+    # CFUNCTYPE,
+    c_ulong,
+    POINTER,
+)
+import sys
+
+
+def get_eof():
+    """ EOF is platform-dependent """
+    stdio_h = open("/usr/include/stdio.h")
+    eof_char = None # Set the correct value here if the script cannot determine it
+    for line in stdio_h:
+        if line.startswith("#define EOF"):
+            eof_char = eval(line.split()[2])
+
+    if eof_char is None:
+        raise Exception("Could not determine the EOF character for this platform. Please set it manually")
+    return eof_char
+
+
+#def get_cuda_signatures(*funcs):
+def get_cuda_signatures():
+    """ Grab CUDA typedefs and signatures from Numba's CUDA driver """
+    signatures = []
+    for func in numba_drvapi.API_PROTOTYPES:
+        ctypes_types = numba_drvapi.API_PROTOTYPES[func] # (ret_type, arg1_type, arg2_type, ...)
+        clisp_types = [get_clisp_type(typ) for typ in ctypes_types]
+        parm_list = [(chr(ord('a') + t), typ) for t, typ in enumerate(clisp_types[1:])]
+        ret_type = clisp_types[0]
+        signatures.append(["define", [[func, ret_type], *parm_list]])
+    globe = globals() # The global scope of this module
+    for memb in dir(numba_drvapi):
+        # All numba_drvapi.cu_* variables are CUDA data types as named by Numba
+        if memb.startswith("cu_"):
+            # Dynamicall create the macro variable
+            globe[memb] = get_clisp_type(getattr(numba_drvapi, memb))
+
+    return signatures
+
+def get_clisp_type(c_type):
+    """ Map a ctypes object to a C-Lisp type """
+    ptr_int8 = ["ptr", "int8"]
+    type_map = {
```

Comment:
I think we can do a better job here. This looks super hacky.

Reply:
I've gone this length for 2 reasons:
* I don't want to inject any code into `numba_drvapi.py`. I want to use it as-is, no modifications, as just another import.
* ctypes seems to create classes on-the-fly. In some cases it is difficult to compare the classes of ctypes objects (e.g. using `isinstance`)

Consider the following Python session:
```python
>>> import numba_drvapi
>>> numba_drvapi.cu_device
<class 'ctypes.c_int'>
>>> import ctypes
>>> isinstance(numba_drvapi.cu_device, ctypes.c_int)
False
```